### ISSUE TITLE
Decimation before ExtractNextContiguousLineSegment

### ISSUE THREAD

**Issue by TSerious** [oxyplot/oxyplot#1748]
### Feature description

This feature helps to improve the performance of LineSeries.

Currently right before the points of a LineSeries are rendered the function ExtractNextContiguousLineSegment is called which, extracts the next line segment and transforms the correspondig DataPoints into ScreenPoints. Afterwards these extracted points can be decimated with the Decimator Action. But the performance impact of the later Decimator is relatively low as ExtractNextContiguousLineSegment is slow.

If a huge number of points shall be plotted but with less detail, it is more performant to do the decimation of of the DataPoints right before the for-loop which calls ExtractNextContiguousLineSegment, as this loop then has to work on less points.

Use case example: The user wants to plot a relative continuous LineSeries with 1e6 points. This exceeds even the resolution of a typical monitor. In this case it would be sufficient to only render for example 500 points.

Implementation suggestion:
- Add a public property `public Func<IList<DataPoint>, double, double, IList<DataPoint>> DataPointDecimator { get; set; }` to `OxyPlot.Series.LineSeries`
- Call that function right before extracting contiguous line segments
```C#
if(DataPointDecimator != null)
{
  points = this.DataPointDecimator(points, XAxis.ActualMinimum, XAxis.ActualMaximum);
  startIdx = 0;
}

for (int i = startIdx; i < points.Count; i++)
{
  if (!this.ExtractNextContiguousLineSegment(points, ref i, ref lastValidPoint, xmax, broken, this.contiguousScreenPointsBuffer))
  {
    break;
  }
  ...
```
- The implementation of the DataPointDecimator function can be done externaly and might be dependant on the use case.

![image](https://user-images.githubusercontent.com/5380109/108999059-12f8b200-76a2-11eb-948f-11e5026d1006.png)
Sine curve with 9000 DataPoints, some NaN values in the middle and without the suggested feature.

![image](https://user-images.githubusercontent.com/5380109/108999423-91555400-76a2-11eb-849a-1c744016a625.png)
Sine curve with 9000 DataPoints, some NaN values in the middle and with the suggested feature (only 60 DatePoints are shown). Areas where points are missing are marked in red.

![image](https://user-images.githubusercontent.com/5380109/108999557-c1045c00-76a2-11eb-8e3e-728229f4696a.png)
Spik sine curve with 9000 DataPoints, some NaN values in the middle and without the suggested feature.

![image](https://user-images.githubusercontent.com/5380109/108999691-f14bfa80-76a2-11eb-85c4-0fc3cc915e08.png)
Spiky sine curve with 9000 DataPoints, some NaN values in the middle and with the suggested feature (only 60 DatePoints are shown).

![image](https://user-images.githubusercontent.com/5380109/108999912-47b93900-76a3-11eb-9bc5-eeca7d220f7e.png)
Zoomed in curve of the previous image. As the user zooms in the details of the graph can be seen and the spike is shown in its correct shape.

![image](https://user-images.githubusercontent.com/5380109/108999766-12145000-76a3-11eb-83b3-4e7394914e8d.png)
Spiky sine curve with 9000 DataPoints, some NaN values in the middle and with the suggested feature where only 500 DatePoints are shown.

**Comment by VisualMelon** [comment]
A pre-transform decimation could be a good idea: have you tried this? We'd have to think about where the responsibility for filtering lies. If you want to put a implementation together, I'll be happy to review it.

Regarding the decimator, if it has many parameters then it should probably be a named delegate, and it should take a list into which it stuffs points, rather than returning a list (as per the existing decimator).

**Comment by TSerious** [comment]
Yes, I've tried this and have achieved good performance results. In a test case with 200 Series (90000 Pts each) I was able to achiev 3 to 4 frames per second, with decimation of the DataPoints to 500 points. Without the decimation I got less than 1 frame per second.

In my (specialized) implementation I only do this decimation when the user pans the graph as this is a situation when you need high or at least a fluent fps. But this might be something that each user has to implement on his/her own, as it also needs binding to the mouse interactions. But I can put together an example for this case as well.

Anyway I also thought about implementing this decimator the same way it is done with the ScreenPoints but this would have resulted in an additional buffer/List (where the points are stuffed into) that must be handled by the LineSeries - just like with the current decimator. That was why I've implemented it like shown, where the List of DataPoints is the return value. But of course I can implement in differently.

**Comment by VisualMelon** [comment]
In your tests, did you disable the monotonicity optimisations, or are you locating the start/end index inside the decimator?

**Comment by TSerious** [comment]
I havn't set IsXMonotonic to false (if this is what you mean) and searched for the stard and end index inside the decimator.
